internal/configv2/configv2.go

```diff
@@ -0,0 +1,511 @@
+
+// processorsKeyName is the configuration key name for processors section.
+const processorsKeyName = "processors"
+
+// pipelinesKeyName is the configuration key name for pipelines section.
+const pipelinesKeyName = "pipelines"
+
+// typeAndNameSeparator is the separator that is used between type and name in type/name composite keys.
+const typeAndNameSeparator = "/"
+
+const (
+	tracesDataTypeStr  = "traces"
+	metricsDataTypeStr = "metrics"
+)
+
+// Load loads a ConfigV2 from Viper.
+func Load(v *viper.Viper) (*configmodels.ConfigV2, error) {
+
+	var config configmodels.ConfigV2
+
+	// Load the config.
+
+	receivers, err := loadReceivers(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Receivers = receivers
+
+	exporters, err := loadExporters(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Exporters = exporters
+
+	processors, err := loadProcessors(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Processors = processors
+
+	pipelines, err := loadPipelines(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Pipelines = pipelines
+
+	// Config is loaded. Now validate it.
+
+	if err := validateConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+// decodeTypeAndName decodes a key in type[/name] format into type and fullName.
+// fullName is the key normalized such that type and name components have spaces trimmed.
+// The "type" part must be present, the forward slash and "name" are optional.
+func decodeTypeAndName(key string) (typeStr, fullName string, err error) {
+	items := strings.SplitN(key, typeAndNameSeparator, 2)
+	if len(items) < 1 {
+		// We need at least one part (the type).
+		err = errors.New("Invalid key")
+		return
+	}
+
+	typeStr = strings.TrimSpace(items[0])
+	if typeStr == "" {
+		// Type must be present.
+		err = errors.New("Invalid key")
+		return
+	}
+
+	var nameSuffix string
+	if len(items) > 1 {
+		// "name" part is present.
+		nameSuffix = strings.TrimSpace(items[1])
+		if nameSuffix == "" {
+			err = errors.New("Invalid key")
+			return
+		}
+	} else {
+		nameSuffix = ""
+	}
+
+	// Create normalized fullName.
+	if nameSuffix == "" {
+		fullName = typeStr
+	} else {
+		fullName = typeStr + typeAndNameSeparator + nameSuffix
+	}
+
+	err = nil
+	return
+}
+
+func loadReceivers(v *viper.Viper) (configmodels.Receivers, error) {
+	// Get the list of all "receivers" sub vipers from config source.
+	subViper := v.Sub(receiversKeyName)
+
+	// Get the map of "receivers" sub-keys.
+	keyMap := v.GetStringMap(receiversKeyName)
+
+	// Prepare resulting map
+	receivers := make(configmodels.Receivers, 0)
+
+	// Iterate over input map and create a config for each.
+	for key := range keyMap {
+		// Decode the key into type and fullName components.
+		typeStr, fullName, err := decodeTypeAndName(key)
+		if err != nil || typeStr == "" {
+			return nil, &configError{
+				code: errInvalidTypeAndNameKey,
+				msg:  fmt.Sprintf("invalid key %q. Key must in the form type[/name]", key),
```

Comment:
if you enhance the errors as mentioned above, it seems worth to include err in the message. Consider also the case that someone later may add some other error message to decodeTypeAndName

Notice that there is no case in which `typeStr == ""` and `err == nil`

Reply:
Added to error message.